### PR TITLE
Enterprise instance uses gateway for attribution

### PR DESCRIPTION
**Pull request by cbart**
Closes [#59262](https://github.com/sourcegraph/sourcegraph/issues/59262)

Enterprise client calls gateway for attribution.

This pull request also:

- moves gateway attribution request/response to common package so that client can use it
- adds tests

This pull request leaves for later to:

- remove all uses of `totalCount` which is not needed
- add configuration for gateway endpoint outside of completions or chat

## Test plan

- Unit testing guardrails client against a fake http server,
- unit testing service backing enterprise resolver against fake guardrails

### CODE REVIEW

File: internal/codygateway/client.go [hunk 1]
```diff
@@ -47,6 +54,7 @@ func NewClientFromSiteConfig(cli httpcli.Doer) (_ Client, ok bool) {
 		return nil, false
 	}
 
+	// TODO: What if customer is BYOK? How do we talk to gateway then?
```
Comment (rafax) [inline internal/codygateway/client.go:57]: _Technically_, you can configure a licence to allow Cody Gateway access even for BYOK users, but we don't have any way for expressing "allow Guardrails calls, but not LLM calls" without resorting to hacks like setting allowedModels to an empty list.

File: internal/codygateway/client.go [hunk 2]
```diff
@@ -129,3 +137,44 @@ func (c *client) GetLimits(ctx context.Context) ([]LimitStatus, error) {
...
 }
+
+func (c *client) Attribution(ctx context.Context, snippet string, limit int) (Attribution, error) {
+	u, err := url.Parse(c.endpoint)
+	if err != nil {
+		return Attribution{}, err
+	}
+	u.Path = "v1/attribution"
+	body := new(bytes.Buffer)
+	// TODO: Use the same struct as gateway
```
[the remark this answers is not in the document]
Comment (rafax) [reply internal/codygateway/client.go:148]: But more seriously, you now have `codygateway.AttributionRequest` available for import - can you use it here?

File: internal/codygateway/client.go (hunk 1 above)
Comment (cbart) [reply internal/codygateway/client.go:57]: yep! I'll add a separate config for guardrails only access https://github.com/sourcegraph/sourcegraph/issues/59576

File: internal/codygateway/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbart) [reply internal/codygateway/client.go]: This actually is that struct - it's the same package. I forgot to remove TODO. Good spot!

**Comment by cbart** [comment]
Thanks @rafax! Please take another look.

**Review by arafatkatze — APPROVED** [review]
Unit Tests work jolly and the PR looks okay. I haven't done a super deep dive though.

**Review by rafax — APPROVED** [review]

**Pull request merged**